build/generator.go

```diff
@@ -27,75 +26,66 @@ type Package struct {
 	Version string
 }
 
-func GetPublishedPackages(c *http.Client, names []string) ([]Package, error) {
+func GetPublishedPackages(c *http.Client, names []string, arch string) ([]Package, error) {
 	var pkgs []Package
 
 	for _, pkgName := range names {
 		ctx, cancelFn := context.WithTimeout(context.Background(), requestTimeout)
 		defer cancelFn()
 
-		url := fmt.Sprintf("%s/%s/+archive/primary", launchpadAPIBaseURL, distroName)
-
-		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, apiBaseURL, nil)
 		if err != nil {
 			return nil, fmt.Errorf("request creation: %w", err)
 		}
 
 		q := req.URL.Query()
-		q.Add("ws.op", "getPublishedBinaries")
-		q.Add("binary_name", pkgName)
-		q.Add("exact_match", "true")
-		q.Add("distro_arch_series", fmt.Sprintf("%s/%s/%s/%s", launchpadAPIBaseURL, distroName, distroVersion, distroArch))
-		q.Add("status", "Published")
-		q.Add("order_by_date", "true")
+		q.Add("keywords", pkgName)
+		q.Add("searchon", "names")
+		q.Add("exact", "1")
+		q.Add("suite", suiteVersion)
+		q.Add("section", "all")
+		q.Add("arch", arch)
 		req.URL.RawQuery = q.Encode()
 
 		resp, err := c.Do(req)
 		if err != nil {
 			return nil, fmt.Errorf("request failed: %w", err)
 		}
 
-		respData := map[string]any{}
-		if err := json.NewDecoder(resp.Body).Decode(&respData); err != nil {
-			return nil, fmt.Errorf("failed to decode response: %w", err)
-		}
-
-		entries, ok := respData["entries"].([]any)
-		if !ok {
-			return nil, fmt.Errorf("failed to parse response")
-		}
-
-		if len(entries) == 0 {
-			break
-		}
-
-		entry, ok := entries[0].(map[string]interface{})
-		if !ok {
-			return nil, fmt.Errorf("failed to parse entry")
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read response body: %w", err)
 		}
 
-		name, ok := entry["binary_package_name"].(string)
-		if !ok {
-			return nil, fmt.Errorf("failed to parse package name")
+		var version string
+		suffixes := []string{fmt.Sprintf(": %s", arch), ": all"}
+		lines := strings.Split(string(data), "\n")
+		for _, line := range lines {
+			if strings.HasSuffix(line, suffixes[0]) || strings.HasSuffix(line, suffixes[1]) {
+				line = strings.TrimSpace(line)
+				version = strings.TrimPrefix(line, "<br>")
+				version = strings.TrimSuffix(version, suffixes[0])
+				version = strings.TrimSuffix(version, suffixes[1])
+				break
+			}
```

Comment:
Unfortunately Debian doesn't seem to offer an API for packages (only for sources) so I literally had to fetch the HTML and do some ugly string matching. I expect this to break at some point but it's working so far so I am happy with it.

Reply:
this can be parsed http://ftp.debian.org/debian/dists/bullseye/main/binary-amd64/Packages.gz

It's not signed but its hash can be found here: http://ftp.debian.org/debian/dists/bullseye/InRelease

Older than newest packages can found at http://ftp.debian.org/debian/dists/bullseye/InRelease

(this can be useful here: https://github.com/paultag/go-debian)

Reply:
@oh6hay Thanks, those indeed look more promising that parsing a raw HTML response :) I'll take a look.